### PR TITLE
Added DBSCANFastRescan

### PR DESCRIPTION
**Pull request by jnpark3**
Implemented DBSCANFastRescan, a class that helps with accelerating DBSCAN hyperparameter search.

### CODE REVIEW

**Comment by jnpark3** [comment]
pre-commit.ci autofix

File: src/gnn_tracking/postprocessing/fastrescanner.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            self._reset_graph(eps)
+
+        filtered_edges = self.edges[self.distances <= eps]
+
+        n_neighbors = np.bincount(filtered_edges[:,0])
+        core_samples = np.asarray(n_neighbors >= min_pts, dtype=np.uint8)
+
+        idx = np.where(np.diff(filtered_edges[:,0]) != 0)[0] + 1
+        n = np.split(filtered_edges[:,1], idx)
+        neighborhoods = np.empty(len(n), dtype=object)
```
Comment (klieret) [inline src/gnn_tracking/postprocessing/fastrescanner.py:55]: Just do `neighborhoods = np.array(np.split(filtered_edges[:,1], idx), dtype="object")`

File: src/gnn_tracking/postprocessing/fastrescanner.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        """
+        neighbors_model = NearestNeighbors(radius=max_eps)
+        neighbors_model.fit(self.X)
+        distances, indexes = neighbors_model.radius_neighbors(self.X, radius=max_eps, return_distance=True, n_jobs=self.n_jobs)
+
+        lengths = [len(arr) for arr in indexes]
+        source_indexes = np.repeat(np.arange(len(indexes)), lengths)
+        target_indexes = np.concatenate(indexes)
+
+        self.distances = np.concatenate(distances)
```
Comment (klieret) [inline src/gnn_tracking/postprocessing/fastrescanner.py:35]: all attributes should be declared in `__init__`. Also mark as private (start with `_`)

File: src/gnn_tracking/postprocessing/fastrescanner.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+        Args:
+            X: Data to cluster
+            max_eps: Maximum epsilon to use during rescanning. Set to as low
+                as possible to save time.
+            n_jobs : int, default=None
+                The number of parallel jobs to run.
+        """
+        self.X = X
+        self.max_eps = max_eps
```
Comment (klieret) [inline src/gnn_tracking/postprocessing/fastrescanner.py:19]: mark as private

File: src/gnn_tracking/postprocessing/fastrescanner.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
+import numpy as np
+from sklearn.neighbors import NearestNeighbors
+from sklearn.cluster._dbscan_inner import dbscan_inner
+from sklearn.cluster import DBSCAN
+
+class DBSCANFastRescan:
+
+    def __init__(self, X, max_eps=1, n_jobs=None):
```
Comment (klieret) [inline src/gnn_tracking/postprocessing/fastrescanner.py:8]: put `, *, ` before `n_jobs`

**Review by klieret — CHANGES_REQUESTED** [review]

**Comment by klieret** [comment]
Btw, press `a` to hide all the annotations by codecov (I've disabled them but that's only gonna affect the next PR)

File: src/gnn_tracking/postprocessing/fastrescanner.py (hunk 4 above)
Comment (klieret) [inline src/gnn_tracking/postprocessing/fastrescanner.py:8]: `X` should be lower case

**Review by klieret — CHANGES_REQUESTED** [review]

**Comment by klieret** [comment]
Please also add [type hints](https://www.pythontutorial.net/python-basics/python-type-hints/)

Also make sure that you have pre-commit installed locally, the autofix didn't seem to work. 

```
pre-commit install
pre-commit run --all
```

**Comment by jnpark3** [comment]
Only change not added was the implementation of neighborhoods = np.array(np.split(filtered_edges[:,1], idx), dtype="object"), as for some reason tests on a notebook showed that this generated a different list of neighborhoods compared to the original implementation

**Comment by jnpark3** [comment]
Only change not added was the implementation of `neighborhoods = np.array(np.split(filtered_edges[:,1], idx), dtype="object")`, as for some reason tests on a notebook showed that this generated a different list of neighborhoods compared to the original implementation

**Comment by klieret** [comment]
Closing this in favor of #378 (which is the same but pushed to a branch)

**Pull request closed without merging**